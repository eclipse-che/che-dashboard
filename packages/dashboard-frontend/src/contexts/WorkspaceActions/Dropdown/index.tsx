/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { helpers } from '@eclipse-che/common';
import {
  AlertVariant,
  Dropdown,
  DropdownItem,
  DropdownProps,
  DropdownToggle,
  KebabToggle,
} from '@patternfly/react-core';
import { CaretDownIcon } from '@patternfly/react-icons';
import React from 'react';

import { ActionContextType } from '@/contexts/WorkspaceActions';
import styles from '@/contexts/WorkspaceActions/Dropdown/index.module.css';
import { lazyInject } from '@/inversify.config';
import { AppAlerts } from '@/services/alerts/appAlerts';
import getRandomString from '@/services/helpers/random';
import { DevWorkspaceStatus, WorkspaceAction, WorkspaceStatus } from '@/services/helpers/types';
import { Workspace } from '@/services/workspace-adapter';

export type Props = {
  context: ActionContextType;
  isDisabled?: boolean;
  toggle: 'kebab-toggle' | 'dropdown-toggle';
  workspace: Workspace;
  onAction?: (
    action: WorkspaceAction,
    workspaceUID: string,
    // true if the action succeeded, false if it failed, undefined if the action was not performed
    succeeded: boolean | undefined,
  ) => Promise<void>;
} & Pick<DropdownProps, 'menuAppendTo' | 'position' | 'isPlain'>;

export type State = {
  isExpanded: boolean;
};

export class WorkspaceActionsDropdown extends React.PureComponent<Props, State> {
  @lazyInject(AppAlerts)
  private appAlerts: AppAlerts;

  constructor(props: Props) {
    super(props);

    this.state = {
      isExpanded: false,
    };
  }

  private buildToggle(): React.ReactElement {
    const { isDisabled = false, toggle, workspace } = this.props;

    if (toggle === 'kebab-toggle') {
      return (
        <KebabToggle
          aria-label="Actions"
          data-testid={`${workspace.uid}-action-dropdown`}
          data-testtype="kebab-toggle"
          isDisabled={isDisabled}
          onToggle={isExpanded => this.handleToggle(isExpanded)}
        />
      );
    }

    return (
      <DropdownToggle
        aria-label="Actions"
        data-testid={`${workspace.uid}-action-dropdown`}
        data-testtype="dropdown-toggle"
        isDisabled={isDisabled}
        onToggle={isExpanded => this.handleToggle(isExpanded)}
        toggleIndicator={CaretDownIcon}
        toggleVariant="primary"
      >
        Actions
      </DropdownToggle>
    );
  }

  private handleToggle(isExpanded: boolean): void {
    this.setState({ isExpanded });
  }

  private async handleSelect(action: WorkspaceAction): Promise<void> {
    const { context, onAction, workspace } = this.props;

    this.handleToggle(false);

    let succeeded: boolean | undefined = undefined;
    try {
      if (action === WorkspaceAction.DELETE_WORKSPACE) {
        try {
          await context.showConfirmation([workspace.name]);
        } catch (e) {
          return onAction?.(action, workspace.uid, undefined);
        }
      }

      await context.handleAction(action, workspace.uid);
      succeeded = true;
    } catch (e) {
      const errorMessage = `Unable to ${action.toLocaleLowerCase()} ${
        workspace.name
      }. ${helpers.errors.getMessage(e)}`;
      this.showAlert(errorMessage);
      console.error(errorMessage);
      succeeded = false;
    }

    return onAction?.(action, workspace.uid, succeeded);
  }

  private showAlert(message: string): void {
    this.appAlerts.showAlert({
      key: 'workspace-dropdown-action-' + getRandomString(4),
      title: message,
      variant: AlertVariant.warning,
    });
  }

  private getDropdownItems(): React.ReactNode[] {
    const { workspace } = this.props;
    const isStopped =
      workspace.status === WorkspaceStatus.STOPPED ||
      workspace.status === DevWorkspaceStatus.STOPPED ||
      workspace.status === DevWorkspaceStatus.FAILED;
    const isTerminating = workspace.status === DevWorkspaceStatus.TERMINATING;

    const getItem = (action: WorkspaceAction, isDisabled: boolean) => {
      return (
        <DropdownItem
          aria-label={`Action: ${action}`}
          component="button"
          key={`action-${action}`}
          isDisabled={isDisabled}
          onClick={async e => {
            e.preventDefault();
            e.stopPropagation();

            return this.handleSelect(action);
          }}
        >
          {action}
        </DropdownItem>
      );
    };

    return [
      getItem(WorkspaceAction.OPEN_IDE, isTerminating),
      getItem(WorkspaceAction.START_DEBUG_AND_OPEN_LOGS, isTerminating || isStopped === false),
      getItem(WorkspaceAction.START_IN_BACKGROUND, isTerminating || isStopped === false),
      getItem(WorkspaceAction.RESTART_WORKSPACE, isTerminating || isStopped),
      getItem(WorkspaceAction.STOP_WORKSPACE, isTerminating || isStopped),
      getItem(WorkspaceAction.DELETE_WORKSPACE, isTerminating),
    ];
  }

  render(): React.ReactElement {
    const { isExpanded } = this.state;
    const { isPlain = false, menuAppendTo = 'inline', position = 'left' } = this.props;

    const dropdownToggle = this.buildToggle();
    const dropdownItems = this.getDropdownItems();

    return (
      <Dropdown
        className={styles.workspaceActionSelector}
        dropdownItems={dropdownItems}
        isOpen={isExpanded}
        isPlain={isPlain}
        menuAppendTo={menuAppendTo}
        position={position}
        toggle={dropdownToggle}
      />
    );
  }
}
