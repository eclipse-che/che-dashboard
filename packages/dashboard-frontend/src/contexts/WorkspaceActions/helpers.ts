/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { WantDelete } from '@/contexts/WorkspaceActions/index';
import { WorkspaceStorageType } from '@/services/models/che';
import { Workspace } from '@/services/workspace-adapter';

export function hasDeleteWarning(
  allWorkspaces: Workspace[],
  wantDelete: WantDelete,
  defaultPvcStrategy: WorkspaceStorageType = '',
): boolean {
  const isPerUserStorageType = (workspace: Workspace) => {
    const storageType = workspace.storageType ? workspace.storageType : defaultPvcStrategy;
    return storageType === 'per-user';
  };

  const perUserWorkspaceToDelete = allWorkspaces.filter(
    workspace => wantDelete.indexOf(workspace.name) !== -1 && isPerUserStorageType(workspace),
  );

  const hasPerUserWorkspaceToDelete = perUserWorkspaceToDelete.length > 0;

  const moreThanOneRunningPerUserWorkspaceToDelete =
    perUserWorkspaceToDelete.filter(
      workspace => workspace.isStarting || workspace.isRunning || workspace.isStopping,
    ).length > 1;

  const hasRunningPerUserWorkspace =
    allWorkspaces.find(
      workspace =>
        wantDelete.indexOf(workspace.name) === -1 &&
        isPerUserStorageType(workspace) &&
        (workspace.isStarting || workspace.isRunning || workspace.isStopping),
    ) !== undefined;

  return (
    moreThanOneRunningPerUserWorkspaceToDelete ||
    (hasPerUserWorkspaceToDelete && hasRunningPerUserWorkspace)
  );
}
