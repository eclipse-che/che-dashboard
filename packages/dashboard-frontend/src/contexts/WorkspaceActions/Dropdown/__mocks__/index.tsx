/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/contexts/WorkspaceActions/Dropdown';
import { WorkspaceAction } from '@/services/helpers/types';

export class WorkspaceActionsDropdown extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    const { isDisabled = false, onAction, workspace } = this.props;
    return (
      <div data-testid="workspace-actions-dropdown">
        <button
          data-testid="workspace-actions-dropdown-open-workspace-button"
          disabled={isDisabled}
          onClick={() => onAction?.(WorkspaceAction.OPEN_IDE, workspace.uid, true)}
        >
          {WorkspaceAction.OPEN_IDE}
        </button>
      </div>
    );
  }
}
