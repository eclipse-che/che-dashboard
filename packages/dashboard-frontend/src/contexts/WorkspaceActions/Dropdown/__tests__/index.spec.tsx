/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { StateMock } from '@react-mock/state';
import userEvent, { UserEvent } from '@testing-library/user-event';
import React from 'react';

import { ActionContextType } from '@/contexts/WorkspaceActions';
import { State, WorkspaceActionsDropdown } from '@/contexts/WorkspaceActions/Dropdown';
import { container } from '@/inversify.config';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';
import { AppAlerts } from '@/services/alerts/appAlerts';
import { AlertItem, DevWorkspaceStatus, WorkspaceAction } from '@/services/helpers/types';
import { Workspace } from '@/services/workspace-adapter';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockHandleAction = jest.fn();
const mockShowConfirmation = jest.fn();
const mockOnAction = jest.fn();

describe('WorkspaceActionsDropdown', () => {
  let workspace: Workspace;
  let user: UserEvent;

  beforeEach(() => {
    workspace = {
      status: DevWorkspaceStatus.STOPPED,
      uid: '1234',
      name: 'my-workspace',
    } as Workspace;

    jest.useFakeTimers();

    user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.clearAllTimers();

    mockHandleAction.mockReset();
  });

  describe('kebab toggle', () => {
    test('snapshot', () => {
      const snapshot = createSnapshot(workspace, 'kebab-toggle');
      expect(snapshot.toJSON()).toMatchSnapshot();
    });

    test('open dropdown', async () => {
      renderComponent(workspace, 'kebab-toggle');

      const toggle = screen.queryByRole('button', { name: 'Actions' });

      expect(toggle).toBeTruthy();
      expect(toggle).toHaveAttribute('data-testtype', 'kebab-toggle');

      // dropdown menu is not visible
      expect(screen.queryByRole('menu', { name: 'Actions' })).toBeNull();

      // toggle dropdown
      const kebabToggle = toggle!;
      await user.click(kebabToggle);

      // now the dropdown menu is visible
      expect(screen.queryByRole('menu', { name: 'Actions' })).not.toBeNull();
    });
  });

  describe('dropdown toggle', () => {
    test('snapshot', () => {
      const snapshot = createSnapshot(workspace, 'dropdown-toggle');
      expect(snapshot.toJSON()).toMatchSnapshot();
    });

    test('open dropdown', async () => {
      renderComponent(workspace, 'dropdown-toggle');

      const toggle = screen.queryByRole('button', { name: 'Actions' });

      expect(toggle).not.toBeNull();
      expect(toggle).toHaveAttribute('data-testtype', 'dropdown-toggle');

      // dropdown menu is not visible
      expect(screen.queryByRole('menu', { name: 'Actions' })).toBeNull();

      // toggle dropdown
      const dropdownToggle = toggle!;
      await user.click(dropdownToggle);

      // now the dropdown menu is visible
      expect(screen.queryByRole('menu', { name: 'Actions' })).not.toBeNull();
    });
  });

  describe('actions handling', () => {
    test('action: open', async () => {
      mockHandleAction.mockResolvedValueOnce(undefined);

      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      const actionOpen = screen.queryByRole('menuitem', { name: 'Action: Open' });

      expect(actionOpen).not.toBeNull();

      await user.click(actionOpen!);

      await jest.advanceTimersByTimeAsync(1000);

      expect(mockShowConfirmation).not.toHaveBeenCalled();
      expect(mockHandleAction).toHaveBeenCalledWith(WorkspaceAction.OPEN_IDE, workspace.uid);
      expect(mockOnAction).toHaveBeenCalledWith(
        WorkspaceAction.OPEN_IDE,
        workspace.uid,
        true, // succeeded
      );
    });

    test('action: open in debug mode', async () => {
      mockHandleAction.mockResolvedValueOnce(undefined);

      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      const actionOpenInDebugMode = screen.queryByRole('menuitem', {
        name: 'Action: Open in Debug mode',
      });

      expect(actionOpenInDebugMode).not.toBeNull();

      await user.click(actionOpenInDebugMode!);

      await jest.advanceTimersByTimeAsync(1000);

      expect(mockShowConfirmation).not.toHaveBeenCalled();
      expect(mockHandleAction).toHaveBeenCalledWith(
        WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        workspace.uid,
      );
      expect(mockOnAction).toHaveBeenCalledWith(
        WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        workspace.uid,
        true, // succeeded
      );
    });

    test('action: start in background', async () => {
      mockHandleAction.mockResolvedValueOnce(undefined);

      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      const actionOpenInBackground = screen.getByRole('menuitem', {
        name: 'Action: Start in background',
      });

      expect(actionOpenInBackground).not.toBeNull();

      await user.click(actionOpenInBackground!);

      await jest.advanceTimersByTimeAsync(1000);

      expect(mockShowConfirmation).not.toHaveBeenCalled();
      expect(mockHandleAction).toHaveBeenCalledWith(
        WorkspaceAction.START_IN_BACKGROUND,
        workspace.uid,
      );
      expect(mockOnAction).toHaveBeenCalledWith(
        WorkspaceAction.START_IN_BACKGROUND,
        workspace.uid,
        true, // succeeded
      );
    });

    test('action: restart workspace', async () => {
      mockHandleAction.mockResolvedValueOnce(undefined);

      workspace.status = DevWorkspaceStatus.RUNNING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      const actionRestartWorkspace = screen.getByRole('menuitem', {
        name: 'Action: Restart Workspace',
      });

      expect(actionRestartWorkspace).not.toBeNull();

      await user.click(actionRestartWorkspace!);

      await jest.advanceTimersByTimeAsync(1000);

      expect(mockShowConfirmation).not.toHaveBeenCalled();
      expect(mockHandleAction).toHaveBeenCalledWith(
        WorkspaceAction.RESTART_WORKSPACE,
        workspace.uid,
      );
      expect(mockOnAction).toHaveBeenCalledWith(
        WorkspaceAction.RESTART_WORKSPACE,
        workspace.uid,
        true, // succeeded
      );
    });

    test('action: stop workspace', async () => {
      mockHandleAction.mockResolvedValueOnce(undefined);

      workspace.status = DevWorkspaceStatus.RUNNING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      const actionStopWorkspace = screen.getByRole('menuitem', {
        name: 'Action: Stop Workspace',
      });

      expect(actionStopWorkspace).not.toBeNull();

      await user.click(actionStopWorkspace!);

      await jest.advanceTimersByTimeAsync(1000);

      expect(mockShowConfirmation).not.toHaveBeenCalled();
      expect(mockHandleAction).toHaveBeenCalledWith(WorkspaceAction.STOP_WORKSPACE, workspace.uid);
      expect(mockOnAction).toHaveBeenCalledWith(
        WorkspaceAction.STOP_WORKSPACE,
        workspace.uid,
        true, // succeeded
      );
    });

    describe('action: delete workspace', () => {
      test('confirmation declined', async () => {
        mockHandleAction.mockResolvedValueOnce(undefined);

        // confirmation declined
        mockShowConfirmation.mockRejectedValueOnce(undefined);

        workspace.status = DevWorkspaceStatus.STOPPED;
        renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

        const actionDeleteWorkspace = screen.getByRole('menuitem', {
          name: 'Action: Delete Workspace',
        });

        expect(actionDeleteWorkspace).not.toBeNull();

        await user.click(actionDeleteWorkspace!);

        await jest.advanceTimersByTimeAsync(1000);

        expect(mockShowConfirmation).toHaveBeenCalledTimes(1);
        expect(mockHandleAction).not.toHaveBeenCalled();
        expect(mockOnAction).toHaveBeenCalledWith(
          WorkspaceAction.DELETE_WORKSPACE,
          workspace.uid,
          undefined, // declined
        );
      });

      test('confirmation accepted, deletion succeeded', async () => {
        mockHandleAction.mockResolvedValueOnce(undefined);

        // confirmation accepted
        mockShowConfirmation.mockResolvedValueOnce(undefined);

        workspace.status = DevWorkspaceStatus.STOPPED;
        renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

        const actionDeleteWorkspace = screen.getByRole('menuitem', {
          name: 'Action: Delete Workspace',
        });

        expect(actionDeleteWorkspace).not.toBeNull();

        await user.click(actionDeleteWorkspace!);

        await jest.advanceTimersByTimeAsync(1000);

        expect(mockShowConfirmation).toHaveBeenCalled();
        expect(mockHandleAction).toHaveBeenCalledWith(
          WorkspaceAction.DELETE_WORKSPACE,
          workspace.uid,
        );
        expect(mockOnAction).toHaveBeenCalledWith(
          WorkspaceAction.DELETE_WORKSPACE,
          workspace.uid,
          true, // succeeded
        );
      });

      test('confirmation accepted, deletion failed', async () => {
        // mute console.error
        console.error = jest.fn();

        mockHandleAction.mockRejectedValueOnce(undefined);

        // confirmation accepted
        mockShowConfirmation.mockResolvedValueOnce(undefined);

        workspace.status = DevWorkspaceStatus.STOPPED;
        renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

        const actionDeleteWorkspace = screen.getByRole('menuitem', {
          name: 'Action: Delete Workspace',
        });

        expect(actionDeleteWorkspace).not.toBeNull();

        await user.click(actionDeleteWorkspace!);

        await jest.advanceTimersByTimeAsync(1000);

        expect(mockShowConfirmation).toHaveBeenCalled();
        expect(mockHandleAction).toHaveBeenCalledWith(
          WorkspaceAction.DELETE_WORKSPACE,
          workspace.uid,
        );
        expect(mockOnAction).toHaveBeenCalledWith(
          WorkspaceAction.DELETE_WORKSPACE,
          workspace.uid,
          false, // failed
        );
      });
    });
  });

  describe('actions status', () => {
    test('workspace FAILED', () => {
      workspace.status = DevWorkspaceStatus.FAILED;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
    });

    test('workspace FAILING', () => {
      workspace.status = DevWorkspaceStatus.FAILING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
    });

    test('workspace RUNNING', () => {
      workspace.status = DevWorkspaceStatus.RUNNING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
    });

    test('workspace STARTING', () => {
      workspace.status = DevWorkspaceStatus.STARTING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
    });

    test('workspace STOPPED', () => {
      workspace.status = DevWorkspaceStatus.STOPPED;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
    });

    test('workspace STOPPING', () => {
      workspace.status = DevWorkspaceStatus.STOPPING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'false');
    });

    test('workspace TERMINATING', () => {
      workspace.status = DevWorkspaceStatus.TERMINATING;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.OPEN_IDE }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.START_DEBUG_AND_OPEN_LOGS,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.START_IN_BACKGROUND }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', {
          name: 'Action: ' + WorkspaceAction.RESTART_WORKSPACE,
        }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.STOP_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'true');
      expect(
        screen.getByRole('menuitem', { name: 'Action: ' + WorkspaceAction.DELETE_WORKSPACE }),
      ).toHaveAttribute('aria-disabled', 'true');
    });
  });

  describe('alerts', () => {
    const mockShowAlert = jest.fn();

    beforeEach(() => {
      class MockAppAlerts extends AppAlerts {
        showAlert(alert: AlertItem): void {
          mockShowAlert(alert);
        }
      }

      container.snapshot();
      container.rebind(AppAlerts).to(MockAppAlerts).inSingletonScope();
    });

    afterEach(() => {
      container.restore();
    });

    test('deletion failed', async () => {
      // mute console.error
      console.error = jest.fn();

      // reject promise - deletion failed
      const errorMessage = 'delete failed';
      mockHandleAction.mockReset();
      mockHandleAction.mockRejectedValueOnce(new Error(errorMessage));

      // confirmation accepted
      mockShowConfirmation.mockResolvedValueOnce(undefined);

      workspace.status = DevWorkspaceStatus.STOPPED;
      renderComponent(workspace, 'kebab-toggle', { isExpanded: true });

      const actionDeleteWorkspace = screen.getByRole('menuitem', {
        name: 'Action: Delete Workspace',
      });

      expect(actionDeleteWorkspace).not.toBeNull();

      await user.click(actionDeleteWorkspace!);

      await jest.advanceTimersByTimeAsync(1000);

      expect(mockShowConfirmation).toHaveBeenCalled();
      expect(mockHandleAction).toHaveBeenCalled();
      expect(mockOnAction).toHaveBeenCalledWith(
        WorkspaceAction.DELETE_WORKSPACE,
        workspace.uid,
        false, // failed
      );

      expect(mockShowAlert).toHaveBeenCalledWith({
        key: expect.stringContaining('workspace-dropdown-action'),
        title: `Unable to delete workspace ${workspace.name}. ${errorMessage}`,
        variant: 'warning',
      });
    });
  });
});

function getComponent(
  workspace: Workspace,
  toggle: 'kebab-toggle' | 'dropdown-toggle',
  localState?: Partial<State>,
): React.ReactElement {
  const context: ActionContextType = {
    handleAction: (...args) => mockHandleAction(...args),
    showConfirmation: (...args) => mockShowConfirmation(...args),
    toDelete: [],
  };

  const component = (
    <WorkspaceActionsDropdown
      context={context}
      onAction={mockOnAction}
      workspace={workspace}
      toggle={toggle}
    />
  );

  if (localState) {
    return <StateMock state={localState}>{component}</StateMock>;
  }

  return component;
}
