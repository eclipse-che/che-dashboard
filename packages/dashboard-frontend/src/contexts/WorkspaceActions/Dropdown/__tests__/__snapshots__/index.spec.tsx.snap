// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkspaceActionsDropdown dropdown toggle snapshot 1`] = `
<div
  className="pf-c-dropdown workspaceActionSelector"
  data-ouia-component-id="OUIA-Generated-Dropdown-3"
  data-ouia-component-type="PF4/Dropdown"
  data-ouia-safe={true}
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    aria-label="Actions"
    className="pf-c-dropdown__toggle pf-m-primary"
    data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
    data-ouia-component-type="PF4/DropdownToggle"
    data-ouia-safe={true}
    data-testid="1234-action-dropdown"
    data-testtype="dropdown-toggle"
    disabled={false}
    id="pf-dropdown-toggle-id-3"
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="pf-c-dropdown__toggle-text"
    >
      Actions
    </span>
    <span
      className="pf-c-dropdown__toggle-icon"
    >
      <svg
        aria-hidden={true}
        aria-labelledby={null}
        fill="currentColor"
        height="1em"
        role="img"
        style={
          {
            "verticalAlign": "-0.125em",
          }
        }
        viewBox="0 0 320 512"
        width="1em"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        />
      </svg>
    </span>
  </button>
</div>
`;

exports[`WorkspaceActionsDropdown kebab toggle snapshot 1`] = `
<div
  className="pf-c-dropdown workspaceActionSelector"
  data-ouia-component-id="OUIA-Generated-Dropdown-1"
  data-ouia-component-type="PF4/Dropdown"
  data-ouia-safe={true}
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    aria-label="Actions"
    className="pf-c-dropdown__toggle"
    data-testid="1234-action-dropdown"
    data-testtype="kebab-toggle"
    disabled={false}
    id="pf-dropdown-toggle-id-0"
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <svg
      aria-hidden={true}
      aria-labelledby={null}
      fill="currentColor"
      height="1em"
      role="img"
      style={
        {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 192 512"
      width="1em"
    >
      <path
        d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
      />
    </svg>
  </button>
</div>
`;
